### PR TITLE
feat(fs/git): add support for cloning target repository to the local filesystem

### PR DESCRIPTION
**Pull request by GeorgeMac**
This adds support for having the Git backend clone to local disk.
The motivation is to offload some of the memory pressure for large repositories onto disk.
This way memory doesn't have to scale in proportion to repository size, with the tradeoff that initial clone and fetch times may be increased due to I/O on local disk.

```yaml
storage:
  type: git
  git:
    backend:
      type: filesytem
      path: /path/to/dir # optional: will create a tempdir in systems temp dir location otherwise
```

### CODE REVIEW

File: internal/config/config_test.go [hunk 1]
```diff
@@ -675,6 +675,9 @@ func TestLoad(t *testing.T) {
 				cfg.Storage = StorageConfig{
 					Type: GitStorageType,
 					Git: &Git{
+						Storage: GitStorageConfig{
```
Comment (markphelps) [inline internal/config/config_test.go:678]: nit: I think its a bit repetitive for the config to be `storage.git.storage`, is there another word we could use ? maybe `backend` ? would go well with other config options like `cache.backend` that we already have, where `memory` is an option

wdyt?

Comment (GeorgeMac) [reply internal/config/config_test.go:678]: Yeah I know what you mean. Backend works for me.

File: internal/config/storage.go [hunk 2]
```diff
@@ -186,6 +188,18 @@ func (g *Git) validate() error {
...
+type GitBackendType string
+
+const (
+	GotBackendMemory     = GitBackendType("memory")
+	GitBackendFilesystem = GitBackendType("filesystem")
+)
+
+type GitBackend struct {
+	Type GitBackendType `json:"type,omitempty" mapstructure:"type" yaml:"type,omitempty"`
+	Path string         `json:"path,omitempty" mapstructure:"path" yaml:"path,omitempty"`
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/config/storage.go:200]: I just felt that we've already gone quite nested. I don't think we will have any more that memory or filesystem here either.

File: internal/config/storage.go [hunk 3]
```diff
@@ -186,6 +188,18 @@ func (g *Git) validate() error {
 	return nil
 }
 
+type GitBackendType string
+
+const (
+	GotBackendMemory     = GitBackendType("memory")
```
[the remark this answers is not in the document]
Comment (GeorgeMac) [reply internal/config/storage.go]: Ooops bit of a go rename macro messup

**Review by markphelps — APPROVED** [review]
nice!

**Pull request merged**